src/solc_0.7/catalyst/AssetAttributesRegistry.sol

```diff
@@ -0,0 +1,206 @@
+    /// @param gemIds list of gems ids to set
+    function setCatalyst(
+        uint256 assetId,
+        uint16 catalystId,
+        uint16[] calldata gemIds
+    ) external virtual override {
+        require(msg.sender == _minter || msg.sender == _upgrader, "NOT_AUTHORIZED_MINTER");
+        _setCatalyst(assetId, catalystId, gemIds, _getBlockNumber());
+    }
+
+    /// @notice sets the catalyst and gems for an asset for a given block number, migration contract only
+    /// @param assetId id of the asset
+    /// @param catalystId id of the catalyst to set
+    /// @param gemIds list of gems ids to set
+    /// @param blockNumber block number
+    function setCatalystWithBlockNumber(
+        uint256 assetId,
+        uint16 catalystId,
+        uint16[] calldata gemIds,
+        uint64 blockNumber
+    ) external override {
+        require(msg.sender == migrationContract, "NOT_AUTHORIZED_MIGRATION");
+        _setCatalyst(assetId, catalystId, gemIds, blockNumber);
+    }
+
+    /// @notice adds gems to an existing list of gems of an asset, minter only
+    /// @param assetId id of the asset
+    /// @param gemIds list of gems ids to set
+    function addGems(uint256 assetId, uint16[] calldata gemIds) external virtual override {
+        require(msg.sender == _upgrader, "NOT_AUTHORIZED_UPGRADER");
+        require(assetId & IS_NFT != 0, "INVALID_NOT_NFT");
+        require(gemIds.length != 0, "INVALID_GEMS_0");
+
+        uint16 catalystId = _records[assetId].catalystId;
+        uint16[MAX_NUM_GEMS] memory gemIdsToStore;
+        if (catalystId == 0) {
+            // fallback on collection catalyst
+            uint256 collectionId = _getCollectionId(assetId);
+            catalystId = _records[collectionId].catalystId;
+            if (catalystId != 0) {
+                _records[assetId].catalystId = catalystId;
+                gemIdsToStore = _records[collectionId].gemIds;
+            }
+        } else {
+            gemIdsToStore = _records[assetId].gemIds;
+        }
+
+        require(catalystId != 0, "NO_CATALYST_SET");
+        uint8 j = 0;
+        uint8 i = 0;
+        for (i = 0; i < MAX_NUM_GEMS; i++) {
+            if (j == gemIds.length) {
+                break;
+            }
+            if (gemIdsToStore[i] == 0) {
+                require(gemIds[j] != 0, "INVALID_GEM_ID");
+                gemIdsToStore[i] = gemIds[j];
+                j++;
+            }
+        }
+        uint8 maxGems = _gemsCatalystsRegistry.getMaxGems(catalystId);
+        require(i <= maxGems, "GEMS_TOO_MANY");
+        _records[assetId].gemIds = gemIdsToStore;
+        uint64 blockNumber = _getBlockNumber();
+        emit GemsAdded(assetId, gemIds, blockNumber);
+    }
+
+    /// @notice set the migratcion contract address, admin or migration contract only
+    /// @param _migrationContract address of the migration contract
+    function setMigrationContract(address _migrationContract) external override {
```

Comment:
@wighawag You commented on this previously here: https://github.com/thesandboxgame/sandbox-smart-contracts/pull/217#discussion_r577007388
Let me know if you see an issue needing fixing.